File: scripts/Analysis/make_figures.r
#install.packages("dplyr")
#install.packages("relaimpo")
#install.packages("scales")

library(dplyr)
library(scales)
require(relaimpo)

#Read in CSV
df <- read.csv("/home/kjbaron/Documents/NABATS/intermediate_files/mcrfs_all.csv", header = TRUE)
#df <- filter(df, num_pre > 0)
#df <- filter(df, num_post > 0)



#df <- filter(df, BFS > 0)
#df <- na.omit(df) #remove rows with na values



# ------------------------------------------------------------------------------------------
#  Table 5
# -----------------------------------------------------------------------------------------



#Create a matrix to fill with R^2 values
r2_results <- matrix(ncol=4, nrow=12)
r2_results[1,] <- c("Project", "Y #Bugs", "Y ChgLines", "Y Exp")
i <- 2

#Calculate R^2 for each dependent variable of each project
for (current_project in c("accumulo","bookkeeper","camel","cassandra","cxf","derby","felix","hive","openjpa","pig","wicket"))
{
  
  #Extract project data and standardize indep. & dep. variables
  p_df <- filter(df, project == current_project)
  p_df[c("LOC","CC","churn","experience","BFS","num_bugs")] <- lapply(p_df[c("LOC","CC","churn","experience","BFS","num_bugs")], function(x) c(scale(x, center= min(x), scale=diff(range(x)))))
  
  #Build generalized linear models for each dependent variable
  m_exp <- glm(formula = experience ~ LOC+CC+churn, data=p_df)
  m_bfs <- glm(formula = BFS ~ LOC+CC+churn, data=p_df)
  m_bugs <- glm(formula = num_bugs ~ LOC+CC+churn, data=p_df)
  
  #Calculate R^2
  #lmg is the R^2 contribution averaged over orderings among regressors
  r2_exp <- calc.relimp(m_exp, type="lmg", rela=FALSE)@R2
  r2_bfs <- calc.relimp(m_bfs, type="lmg", rela=FALSE)@R2
  r2_bugs <- calc.relimp(m_bugs, type="lmg", rela=FALSE)@R2
  
  #Add results to maxtrix
  r2_results[i,] <- c(current_project, r2_exp, r2_bfs, r2_bugs)
  i <- i+1
}



#  The code below is mostly the same as for Table 5, but rela=TRUE in the relimp calculation 
#  -- just separated the code by table to keep it easy to read
#  --------------------------------------------------------------------------------------------------------------------------
#  TABLE 6 - Average contributions from each independent variable when different dependent variables are used
#  --------------------------------------------------------------------------------------------------------------------------



#Create a matrix to fill with values
t6_results <- matrix(ncol=5, nrow=34)
t6_results[1,] <- c("Project", "Features", "Y #Bugs", "Y ChgLines", "Y Exp")
i <- 2

#Calculate contributions for each dependent variable of each project
for (current_project in c("accumulo","bookkeeper","camel","cassandra","cxf","derby","felix","hive","openjpa","pig","wicket"))
{
  
  #Extract project data and standardize indep. & dep. variables
  p_df <- filter(df, project == current_project)
  p_df[c("LOC","CC","churn","experience","BFS","num_bugs")] <- lapply(p_df[c("LOC","CC","churn","experience","BFS","num_bugs")], function(x) c(scale(x, center= min(x), scale=diff(range(x)))))
  
  #Build generalized linear models for each dependent variable
  m_exp <- glm(formula = experience ~ LOC+CC+churn, data=p_df)
  m_bfs <- glm(formula = BFS ~ LOC+CC+churn, data=p_df)
  m_bugs <- glm(formula = num_bugs ~ LOC+CC+churn, data=p_df)
  
  #Calculate R^2
  #lmg is the R^2 contribution averaged over orderings among regressors
  lmg_exp <- calc.relimp(m_exp, type="lmg", rela=TRUE)@lmg
  lmg_bfs <- calc.relimp(m_bfs, type="lmg", rela=TRUE)@lmg
  lmg_bugs <- calc.relimp(m_bugs, type="lmg", rela=TRUE)@lmg
  
  
  #Add results to maxtrix
  t6_results[i,] <- c(current_project, "X LOC", lmg_bugs['LOC'], lmg_bfs['LOC'], lmg_exp['LOC'])
  t6_results[i+1,] <- c(current_project, "X CC", lmg_bugs['CC'], lmg_bfs['CC'], lmg_exp['CC'])
  t6_results[i+2,] <- c(current_project, "X churn", lmg_bugs['churn'], lmg_bfs['churn'], lmg_exp['churn'])
  i <- i+3
}

File: scripts/Analysis/make_figures.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import statistics
import os



"""

IMPORTANT NOTE:

Figures for fig2 are a bit different than originals. I counted # bugs for dev exp and counted # files for BFS and cost.

ASK GEMA if I should do all at file level?

"""


dir = "/home/kjbaron/Documents/NABATS/"

mcrfs = pd.read_csv(dir+"major_consecutive_releases_files.csv")
mcrfs = mcrfs.loc[mcrfs['BFS'] > 0]
projects = ["accumulo","bookkeeper","camel","cassandra","cxf","derby","felix","hive","openjpa","pig","wicket"]


def cost_calc(l1, l2):
    cost = []
    for x in range(len(l1)):
        cost.append(l1[x]*l2[x])
    return cost

def figure1():
    bfs = []
    exp = []
    cost = []

    for project in projects:
        bfs_list = mcrfs.loc[mcrfs['project'] == project]['BFS'].tolist()
        exp_list = mcrfs.loc[mcrfs['project'] == project]['mean_experience'].tolist()
        print(min(exp_list))
        bfs.append(bfs_list)
        exp.append(exp_list)
        cost.append(cost_calc(exp_list, bfs_list))

    plt.figure()
    plt.boxplot(bfs)
    plt.yscale('log')
    plt.xticks(range(1,12), projects)
    plt.ylabel("Bug Fix Size")

    plt.figure()
    plt.boxplot(exp)
    plt.yscale('symlog')
    plt.xticks(range(1,12), projects)
    plt.ylabel("Dev Experience")

    plt.figure()
    plt.boxplot(cost)
    plt.yscale('symlog')
    plt.xticks(range(1,12), projects)
    plt.ylabel("Cost")

    plt.show()




def bar_chart(list, x_label, y_label, project):
    counts = []
    num_groups = 10
    scale = int(max(list)/num_groups)
    scale = scale+(10-scale%10)

    if x_label == "Cost":
        scale = 1000


    for x in range(1,11):
        counts.append(len([i for i in list if i > (x-1)*scale and i < (x*scale)]))

    plt.figure()
    #fig, ax = plt.subplots()
    plt.bar(range(num_groups), counts)
    plt.xticks(range(1,num_groups+1), range(scale, int(max(list)), scale))
    plt.ylabel(y_label)
    plt.xlabel(x_label)
    plt.title(project)
    plt.tight_layout()



    plt.savefig(dir+"Figures/Figure2/"+project+'_'+x_label+'.png')



def figure2():
    for project in projects:
        p_df = mcrfs.loc[mcrfs['project'] == project]

        #Exp will be the same for all files so just look for the mean exp from the first occurance of each bug_id
        exp_dict = dict()
        exps = []
        bfss = []
        costs = []
        max_bfs = 100
        max_cost = 10000
        for index, row in p_df.iterrows():
            if row['bug_id'] in exp_dict:
                exp_dict[row['bug_id']].append(row['mean_experience'])
            else:
                exp_dict[row['bug_id']] = [row['mean_experience']]

            if row['BFS'] < max_bfs:
                bfss.append(row['BFS'])

            cost = row['BFS']*row['mean_experience']
            if cost < max_cost:
                costs.append(cost)

        for key in exp_dict:
            exps.append(statistics.mean(exp_dict[key]))


        bar_chart(exps, "Dev_Experience", "#_Bugs", project)
        bar_chart(bfss, "Bug_Fix_Size", "#_Bugs", project)
        bar_chart(costs, "Cost", "#_Bugs", project)


figure1()
